File: packages/server/app/models/transaction.py
from typing import Literal

from pydantic import BaseModel

from .cart import CartItem

PaymentMethod = Literal["preauth", "loyalty", "card"]
PaymentStatus = Literal["pending_capture", "captured", "failed"]


class TransactionPayment(BaseModel):
    method: PaymentMethod
    preauth_token: str | None = None
    amount: float
    status: PaymentStatus = "pending_capture"
    settled_at: str | None = None


class TransactionTotals(BaseModel):
    subtotal: float
    vat_total: float
    total: float


class Transaction(BaseModel):
    type: Literal["transaction"] = "transaction"
    txn_id: str
    cart_id: str
    user_id: str
    store_id: str
    device_id: str
    items: list[CartItem] = []
    totals: TransactionTotals
    payment: TransactionPayment
    points_earned: int
    offline: bool = False
    created_at: str
    synced_at: str | None = None

File: packages/shared/src/models/transaction.ts
import { CartItem } from "./cart";

export type PaymentMethod = "preauth" | "loyalty" | "card";
export type PaymentStatus = "pending_capture" | "captured" | "failed";

export interface TransactionPayment {
  method: PaymentMethod;
  preauth_token: string | null;
  amount: number;
  status: PaymentStatus;
  settled_at: string | null;
}

export interface TransactionTotals {
  subtotal: number;
  vat_total: number;
  total: number;
}

export interface Transaction {
  type: "transaction";
  txn_id: string;
  cart_id: string;
  user_id: string;
  store_id: string;
  device_id: string;
  items: CartItem[];
  totals: TransactionTotals;
  payment: TransactionPayment;
  points_earned: number;
  offline: boolean;
  created_at: string;
  synced_at: string | null;
}
